pkg/adaptor/elasticsearch/clients/v2/writer.go

```diff
@@ -0,0 +1,95 @@
+package v2
+
+import (
+	"sync"
+	"time"
+
+	elastic "gopkg.in/olivere/elastic.v3"
+
+	"github.com/compose/transporter/pkg/adaptor/elasticsearch/clients"
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+	version "github.com/hashicorp/go-version"
+)
+
+var (
+	_ client.Writer  = &Writer{}
+	_ client.Session = &Writer{}
+)
+
+// Writer implements client.Writer and client.Session for sending requests to an elasticsearch
+// cluster via its _bulk API.
+type Writer struct {
+	bp *elastic.BulkProcessor
+}
+
+func init() {
+	constraint, _ := version.NewConstraint(">= 2.0, < 5.0")
+	clients.Add("v2", constraint, func(done chan struct{}, wg *sync.WaitGroup, opts *clients.ClientOptions) (client.Writer, error) {
+		esOptions := []elastic.ClientOptionFunc{
+			elastic.SetURL(opts.URLs...),
+			elastic.SetSniff(false),
+			elastic.SetHttpClient(opts.HTTPClient),
+			elastic.SetMaxRetries(2),
+			elastic.SetInfoLog(log.Base().With("path", opts.Path)),
+		}
+		if opts.UserInfo != nil {
+			if pwd, ok := opts.UserInfo.Password(); ok {
+				esOptions = append(esOptions, elastic.SetBasicAuth(opts.UserInfo.Username(), pwd))
+			}
+		}
+		esClient, err := elastic.NewClient(esOptions...)
+		if err != nil {
+			return nil, err
+		}
+		return newWriter(esClient, done, wg), nil
+	})
+}
+
+func newWriter(client *elastic.Client, done chan struct{}, wg *sync.WaitGroup) *Writer {
+	p, _ := client.BulkProcessor().
```

Comment:
Not sure what we'd do with the errors, but right now we're not catching any of the errors from the BulkProcessor. You can use a [BulkAfterFunc](https://github.com/olivere/elastic/blob/1368b91516f39100baf5e23a02ae9f090c0d90e2/bulk_processor.go#L70) and [.After(callback)](https://github.com/olivere/elastic/blob/1368b91516f39100baf5e23a02ae9f090c0d90e2/bulk_processor.go#L78) to get at the errors though. Would be bad to throw away errors without any output.